[unicorn] Add authentication + transactions log

mafesan opened the pull request:
About authentication (from issue #35):

After this change, users can perform authenticated queries to Bestiary. For doing so, they must generate a JWT (JSON Web Token) which has to be included within the `Authorization` header in the HTTP requests. 

About transactions log (from issue #36): 

This change includes methods for generating, verifying and refreshing these tokens.

The purpose of this feature is to store in a log each successful operation run by Bestiary adding, deleting or modifying any entity in the database.

---

Note: these two features come together because they couldn't be implemented separately. When trying to implement only the three mutations for setting up authentication coming from graphql_jwt module, it caused the GraphQL endpoint to fail, as there are no queries to show.

sduenas commented on config/settings/testing.py:
```diff
@@ -0,0 +1,63 @@
+import sys
+import logging
```
This should be:

```suggestion
import logging
import sys
```

sduenas commented on releases/unreleased/support-authenticated-queries.yml:
```diff
@@ -0,0 +1,14 @@
+---
+title: Support authenticated queries
```
I would call it `Authenticated queries`

sduenas commented on releases/unreleased/support-authenticated-queries.yml:
```diff
@@ -0,0 +1,14 @@
+    queries to Bestiary. For doing so, they must
+    generate a JWT (JSON Web Token) which has to be
+    included within the `Authorization` header in the
+    HTTP requests. 
```
This text shouldn't be like a commit message, so don't start it with `After this change`. I would be better something like `Users will be able to perform authenticated queries to Bestiary` or `This feature allows to perform authenticated queries to Bestiary`.

sduenas commented on releases/unreleased/support-authenticated-queries.yml:
```diff
@@ -0,0 +1,14 @@
+    HTTP requests. 
+
+    This change includes methods for generating,
+    verifying and refreshing these tokens.
```
Don't talk about `this change`. You also should explain which methods were added and how to use them.

sduenas commented on bestiary/core/errors.py:
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python
```
Remove this line. This is not a script.

mafesan replied:
I have removed this line from the rest of the tests too.

sduenas commented on bestiary/core/errors.py:
```diff
@@ -0,0 +1,73 @@
+CODE_BASE_ERROR = 1
+CODE_ALREADY_EXISTS_ERROR = 2
+CODE_NOT_FOUND_ERROR = 3
+CODE_VALUE_ERROR = 4
```
It would make more sense to define this values with `Enum`. See an example here: https://github.com/chaoss/grimoirelab-kingarthur/blob/master/arthur/tasks.py#L55

sduenas commented on bestiary/core/models.py:
```diff
@@ -0,0 +1,96 @@
+MAX_SIZE_CHAR_FIELD = 128
+
+
+class Transaction(Model):
```
I think it's a good moment to start adding documentation to the model.

sduenas commented on bestiary/core/log.py:
```diff
@@ -0,0 +1,170 @@
+# Authors:
+#     Miguel Ángel Fernández <mafesan@bitergia.com>
+#
+
```
Remove this line

sduenas commented on bestiary/core/log.py:
```diff
@@ -0,0 +1,170 @@
+from .errors import AlreadyExistsError, ClosedTransactionError
+from .models import (Operation,
+                     Transaction)
+
```
Remove this line.

sduenas commented on tests/test_log.py:
```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python
```
Remove this line.

sduenas commented on tests/test_utils.py:
```diff
@@ -0,0 +1,80 @@
+#     Miguel Ángel Fernández <mafesan@bitergia.com>
+#
+
+
```
Remove this line.

sduenas commented on bestiary/core/schema.py:
```diff
@@ -20,10 +20,176 @@
+
+
+class AbstractPaginatedType(graphene.ObjectType):
+
```
I think we should add documentation to the schema too.

mafesan replied:
I have added this piece of documentation, but I was not sure about how detailed should I document the rest of the classes such as `PaginationType`.

sduenas commented on tests/test_schema.py:
```diff
@@ -0,0 +1,587 @@
+#!/usr/bin/env python
```
Remove this line.

sduenas commented on releases/unreleased/add-transactions-log.yml:
```diff
@@ -0,0 +1,9 @@
+---
+title: Add transactions log
```
Rename it to `Transactions log`.

sduenas commented on releases/unreleased/add-transactions-log.yml:
```diff
@@ -0,0 +1,9 @@
+notes: >
+    The purpose of this feature is to store in a log
+    each successful operation run by Bestiary adding,
+    deleting or modifying any entity in the database.
```
You should explain which queries are available to get that data.

mafesan commented on bestiary/core/models.py:
```diff
@@ -0,0 +1,139 @@
+    * closed_at`: Datetime when the transaction is closed.
+    * is_closed`: Boolean value, `True` if the transaction is closed.
+    * authored_by`: Username from who is creating the transaction.
+    """
```
I was not sure if this was the right way to document Model classes. @sduenas What do you think about the format above?

mafesan commented:
Hi @sduenas, I have addressed your comments. I have also added docuementation in `schema.py` which is displayed in `GraphiQL` interface, for the future consumers of this API. Thanks!

sduenas commented on bestiary/core/models.py:
```diff
@@ -0,0 +1,139 @@
+
+
+class Transaction(Model):
+    """Model class for Transaction objects
```
I think you forgot to add a period here.

sduenas commented on bestiary/core/models.py:
```diff
@@ -0,0 +1,151 @@
+
+
+class Operation(Model):
+    """Model class for Transaction objects
```
We need a period in here.

sduenas commented on releases/unreleased/authenticated-queries.yml:
```diff
@@ -0,0 +1,19 @@
+    * One for creating new tokens, taking as input an existing
+    username and its password.
+    * Another one for verifying existing tokens
+    * Another one for refreshing a token from an existing one.
```
You should enumerate the name of the methods here.

sduenas commented on releases/unreleased/transactions-log.yml:
```diff
@@ -0,0 +1,22 @@
+    The query getting the operation objects includes
+    information like the parent transaction, the operation
+    id, its type, its creation date and its the target
+    entity, among other fields.
```
Add an empty line at the end of the file.

mafesan commented:
@sduenas Oops! I'm sorry about the style mistakes. I have just addressed your last comments, thanks a lot!
